README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# viscomplexr

<!-- badges: start -->

[![R build
status](https://github.com/PeterBiber/viscomplexr/workflows/R-CMD-check/badge.svg)](https://github.com/PeterBiber/viscomplexr/actions)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/viscomplexr)](https://cran.r-project.org/package=viscomplexr)
<!-- badges: end -->

<!-- Make a pretty banner. The code is executed but not displayed here. Below, it is displayed in the example section, but not exectued -->

<img src="man/figures/README-banner-1.png" width="100%" style="display: block; margin: auto;" />
<br>

With *viscomplexr* you can create phase portraits of functions on the
complex number plane. While the main purpose of this package are
scientific, educational and technical applications, the sheer beauty of
phase portraits may be motivation enough to explore the fascinating
realm of complex-valued functions. The package is embedded in the
framework of R base graphics and allows for print quality graphics
files. To a great deal, *viscomplexr* follows the conventions published
by Wegert (2012).

For a detailed step-by-step introduction see the package’s vignette.
After installing *viscomplexr* on your machine, you can view the
vignette by calling:

``` r
vignette("viscomplexr-vignette")
```

Besides making you acquainted with how to work with the package, the
vignette also tells you more about some shenanigans going on behind the
scenes; among them parallel processing and *viscomplexr*’s temporary
file system.

## Installation

The released version of *viscomplexr* can be installed from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("viscomplexr")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("PeterBiber/viscomplexr")
```

## Example

Let’s plot a phase portrait of the function `f(z)=sin(z^3/(z-1i+2))`.
This function is portrayed in the banner above, and the code below
produces exactly that banner. The workhorse function of *viscomplexr* is
called `phasePortrait`, and besides its specific input variables, it
also accepts all parameters that are accepted by R’s `plot.default`
function, like `axes`, `xaxs`, and `yaxs` in the example below. Other
often useful ones are `main`, `xlab`, and `ylab`, especially for
scientific, technical or educational applications.

``` r
# Make a phase portrait of the function 
# f(z) = sin(z^3/(z - 1i + 2))
# in the domain [-6, 6] (real), and [-2.4, 2.4] (imaginary)

library(viscomplexr)

# You can open any R graphics device here (bmp, jpeg, png, tiff, pdf,
# windows, x11, ...), see R documentation if you are not familiar with these.
# This example, as it is, plots into the default graphics device.

# reduce plot margin to 0, store previous graphics parameter settings in op
op <- par(mar = c(0, 0, 0, 0))
# call the function phasePortrait
phasePortrait("sin(z^3/(z - 1i + 2))",               # define the function
              xlim = c(-6, 6), ylim = c(-2.4, 2.4),  # define the domain
              axes = FALSE,                          # suppress axes
              xaxs = "i", yaxs = "i")          # no empty zone around plot
# set graphics parameters to previous values
par(op)

# If you have initially opened a graphics device that plots into a file, you 
# must close it here with
# dev.off()
```

While there are many options and a few useful supporting functions (see
vignette and documentation), this example shows the core way of plotting
phase portraits. It could even be reduced to just the function to be
portrayed (here provided as a character string) and the parameters
`xlim` and `ylim`. The other parameters in the example, and also the
omission of the plot margins before calling *phasePortrait*, were only
used in order to obtain a pretty banner for this README.

## Acknowledgments

While this package is a leisure project, it would have been a mission
impossible without the background of my (P.B.) daily work with R as a
Forest Scientist at the Technical University of Munich (TUM).
Fortunately, I have a job that allows me to learn about Nature by asking
her questions (or trying to simulate what she is doing) with
ever-improving methods and tools. I would like to thank everyone at the
Chair of Forest Growth and Yield Science at TUM who keep me involved in
discussions like: *How can this be solved in R …*

``` r
switch(1 + trunc(runif(1, 0, 6)),
       "... at all?",
       "... in a quick-and-dirty way?",
       "... in Hadley-Wickham-style?",
       "... without a loop?",
       "... without nested loops?",
       "... in a way somebody can understand?")
```

**Veronika Biber** provided expert advice for improving the vignette.
**Johannes Biber** turned out the most patient pre-release tester one
can imagine, boosting things with his high-end gaming machine. Thanks,
guys\! Also thanks to **Gregor Seyer** for his helpful review of the
CRAN submission.

Clearly, programming in R would not be what it is, weren’t there some R
titans who generously share their knowledge online. While I keep
learning from all of them, I would like to thank especially **Hadley
Wickham** and **Dirk Eddelbüttel**.

## References

<div id="refs" class="references">

<div id="ref-wegert_visualcpx_2012">

Wegert, Elias. 2012. *Visual Complex Functions. An Introduction with
Phase Portraits*. Basel Heidelberg New York Dordrecht London: Springer.

</div>

</div>


NEWS.md
# viscomplexr 1.0.1

Fixed the issues pointed out by Gregor Seyer in his review of the CRAN submission:

* Unwrapped all examples from \\dontrun{}. Most of them had to be wrapped in \\donttest{}, however, due to execution time > 5 sec
* Implemented the option *verbose* in *phasePortrait* which allows to suppress all progress messages
* The environments created in order to allow handling large arrays by pointers do now have the empty environment as their parent (not the global environment)
* Made sure that no example, neither in the documentation of the functions, nor in the vignette, does use more than two processor cores

------
# viscomplexr 1.0.0

* Added a `NEWS.md` file to track changes to the package.
* Submitted version 1.0.0 to CRAN


tests/testthat.R
library(testthat)
library(viscomplexr)

test_check("viscomplexr")


tests/testthat/test-funsWork.R
# -----------------------------------------------------------------------------
# Test if phasePortrait produces the correct numerical output.
#
# Uses reference files created with the R-script in CreateTestCases/.
# These reference files are in the project's subdirectory tests/testthat/,
# which automatically becomes the working directory when test_that() is run.
# -----------------------------------------------------------------------------


# function cleanUp
# Deletes all files from the R session's temporary directory which match the
# naming conventions of phasePortrait's temporary files. Is called before and
# after each test run. Thus, the R session's temporary directory, is always
# absolutely clean and unambiguous as far as phasePortrait output is concerned.
cleanUp <- function() {
  existZWMat <- dir(tempdir(), pattern = "^\\d+(z|w)mat\\d{10}\\.RData$")
  unlink(paste0(tempdir(), "/", existZWMat))
}


# function loadActual
# After a call to cleanup and a subsequent call of phasePortrait, only output
# files from the last phasePortrait run can possibly reside in the R session's
# temporary directory (if we explicitly told phasePortrait not to delete them
# after use). The test cases are so slim, that there is only one zmat and one
# wmat file. We load the "wmat" file for further inspection.
loadActual <- function() {
  existWMat <- dir(tempdir(), pattern = "^\\d+wmat\\d{10}\\.RData$")
  get(load(paste0(tempdir(), "/", existWMat)))
}



# Test cases defined as functions

# Test case 1:
# A rational function given as single string
testCase1 <- function() {
  cleanUp()
  phasePortrait("(2-z)^2*(-1i+z)^3*(4-3i-z)/((2+2i+z)^4)",
                xlim = c(-4, 4), ylim = c(-4, 4),
                invertFlip = FALSE,
                blockSizePx = 2250000,
                res = 150,
                tempDir = NULL,
                deleteTempFiles = FALSE,
                noScreenDevice = TRUE,
                nCores = 2,
                verbose = FALSE)
  referenceWmat <- get(load("1wmatCase001.RData"))
  actualWmat    <- loadActual()
  cleanUp()
  rslt          <- all.equal(referenceWmat, actualWmat)
  rm(referenceWmat, actualWmat)
  return(rslt)
}


# Test case 2:
# A rational function given as single string, but with invertFlip = TRUE
testCase2 <- function() {
  cleanUp()
  phasePortrait("(2-z)^2*(-1i+z)^3*(4-3i-z)/((2+2i+z)^4)",
                xlim = c(-4, 4), ylim = c(-4, 4),
                invertFlip = TRUE,
                blockSizePx = 2250000,
                res = 150,
                tempDir = NULL,
                deleteTempFiles = FALSE,
                noScreenDevice = TRUE,
                nCores = 2,
                verbose = FALSE)
  referenceWmat <- get(load("1wmatCase002.RData"))
  actualWmat    <- loadActual()
  cleanUp()
  rslt          <- all.equal(referenceWmat, actualWmat)
  rm(referenceWmat, actualWmat)
  return(rslt)
}


# Test case 3
# User function with additional default arguments which are _not_ specified
# in the call to phasePortrait
testCase3 <- function() {

  jacobiTheta <- function(z, tau = 1i, nIter = 30) {
    k <- c(1:nIter)
    q <- exp(pi*1i*tau)
    g <- exp(2*pi*1i*z)
    return(1 + sum(q^(k^2)*g^k + q^(k^2)*(1/g)^k))
  }

  cleanUp()
  phasePortrait(jacobiTheta,
                xlim = c(-2, 2), ylim = c(-2, 2),
                invertFlip = FALSE,
                blockSizePx = 2250000,
                res = 150,
                tempDir = NULL,
                deleteTempFiles = FALSE,
                noScreenDevice = TRUE,
                nCores = 2,
                verbose = FALSE)
  referenceWmat <- get(load("1wmatCase003.RData"))
  actualWmat    <- loadActual()
  cleanUp()
  rslt          <- all.equal(referenceWmat, actualWmat)
  rm(referenceWmat, actualWmat)
  return(rslt)
}


# Test case 4
# User function with additional default arguments which are specified
# in the call to phasePortrait
testCase4 <- function() {

  jacobiTheta <- function(z, tau = 1i, nIter = 30) {
    k <- c(1:nIter)
    q <- exp(pi*1i*tau)
    g <- exp(2*pi*1i*z)
    return(1 + sum(q^(k^2)*g^k + q^(k^2)*(1/g)^k))
  }

  cleanUp()
  phasePortrait(jacobiTheta,
                moreArgs = list(tau = 1i/2 - 1/4, nIter = 30),
                xlim = c(-2, 2), ylim = c(-2, 2),
                invertFlip = FALSE,
                blockSizePx = 2250000,
                res = 150,
                tempDir = NULL,
                deleteTempFiles = FALSE,
                noScreenDevice = TRUE,
                nCores = 2,
                verbose = FALSE,
                autoDereg = TRUE) # Register sequential backend after
                                  # the last phase portrait
  referenceWmat <- get(load("1wmatCase004.RData"))
  actualWmat    <- loadActual()
  cleanUp()
  rslt          <- all.equal(referenceWmat, actualWmat)
  rm(referenceWmat, actualWmat)
  return(rslt)
}



# The actual tests
test_that("phasePortrait produces correct numerical output", {
  expect_true(testCase1())
  expect_true(testCase2())
  expect_true(testCase3())
  expect_true(testCase4())
})

# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------




tests/testthat/test-funDefinitions.R
# -----------------------------------------------------------------------------
# Test that phasePortrait works with all different ways of defining
# Input functions and their arguments
# -----------------------------------------------------------------------------

# The function makeFunctionFromInput handles the user input
# if it returns an object which is a function (instead of NULL),
# this function has also be tested for producing a useful
# result with a random complex number (and additional arguments,
# if such are provided by the user).
test_that("All allowed kind of inputs are transformed into
          working functions", {
  # Single string input without more arguments
  out <- makeFunctionFromInput("(2-z)^2*(-1i+z)^3*(4-3i-z)/((2+2i+z)^4)")
  expect_equal(is.function(out), TRUE)
  # Single string input with more arguments
  out <- makeFunctionFromInput("sin(r/z)^(z-b)",
                               moreArgs = list(b = 1.2, r = 1))
  expect_equal(is.function(out), TRUE)
  # Single string input with vapply
  out <- makeFunctionFromInput(
    "vapply(z, function(z, a){
      return(prod(abs(a)/a * (a-z)/(1-Conj(a)*z)))
     },
     a = c(0.12152611+0.06171533i,  0.53730315+0.32797530i,
           0.35269601-0.53259644i, -0.57862039+0.33328986i,
          -0.94623221+0.06869166i, -0.02392968-0.21993132i),
    FUN.VALUE = complex(1))")
  expect_equal(is.function(out), TRUE)
  # Match.fun compatible definition with string
  out <- makeFunctionFromInput("cos")
  expect_equal(is.function(out), TRUE)
  # Match.fun compatible definition with function object
  out <- makeFunctionFromInput(tan)
  expect_equal(is.function(out), TRUE)
  # Match.fun compatible definition with user function object
  whatEver <- function(z) {
    (1.2 - z)^cos(z)
  }
  out <- makeFunctionFromInput(whatEver)
  expect_equal(is.function(out), TRUE)
  # Match.fun compatible definition with user function name
  strangeFun <- function(z) {
    (1.2 - z)/cos(z)
  }
  out <- makeFunctionFromInput("strangeFun")
  expect_equal(is.function(out), TRUE)
  # Match.fun compatible definition with user function object
  # and moreArgs
  jacobiTheta <- function(z, tau = 1i, nIter = 30) {
    k <- c(1:nIter)
    q <- exp(pi*1i*tau)
    g <- exp(2*pi*1i*z)
    return(1 + sum(q^(k^2)*g^k + q^(k^2)*(1/g)^k))
  }
  out <- makeFunctionFromInput(jacobiTheta, moreArgs = list(tau = 1i/2 - 1/4))
  expect_equal(is.function(out), TRUE)
  # Match.fun compatible definition with user function name
  # and moreArgs
  out <- makeFunctionFromInput("jacobiTheta", moreArgs = list(tau = 1i/2 - 1/4))
  expect_equal(is.function(out), TRUE)
  # Match.fun compatible definition with user function object
  # but without moreArgs
  out <- makeFunctionFromInput(jacobiTheta)
  expect_equal(is.function(out), TRUE)
  # Match.fun compatible definition with user function name
  # but without moreArgs
  out <- makeFunctionFromInput("jacobiTheta")
  expect_equal(is.function(out), TRUE)
})

# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------



R/Helpers.R
# R package viscomplexr - phase portraits of functions in the
# complex number plane
# Copyright (C) 2020 Peter Biber
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>



#' Adjust ylim to xlim
#'
#' This simple function is useful for adjusting x and y coordinate ranges
#' \code{xlim} and \code{ylim} in order to maintain a desired display ratio. The
#' former must be given, the latter will be adjusted.
#'
#' For certain purposes, e.g. producing a graph that exactly matches a screen,
#' the x and y coordinates must be adjusted to match a given display ratio. If
#' the horizontal range, \code{xlim}, the desired ratio, \code{x_to_y} and the
#' desired center of the y-range, \code{centerY} are provided, this function
#' returns an adapted vertical range, that can be used as \code{ylim} in any
#' plot including \code{\link{phasePortrait}}.
#'
#' @param xlim Numeric vector of length 2; the fixed lower and upper boundary
#'   of the horizontal coordinate range
#'
#' @param centerY The vertical coordinate which the output range is to be
#'   centered around (default = 0)
#'
#' @param x_to_y The desired ratio of the horizontal (x) to the vertical (y)
#'   range. Default is 16/9, a display ratio frequently used for computer or
#'   mobile screens
#'
#' @return A numeric vector of length 2; the lower and upper boundary of the
#'   resulting vertical coordinate range
#'
#' @family helpers
#'
#' @export
#'
#' @examples
#' # Make a phase portrait of a Jacobi theta function that fully covers a
#' # plot with a display aspect ratio of 4/3.
#' # 10 inch wide window with 4/3 display ratio (x/y)
#' \donttest{
#' # x11(width = 10, height = 10 * 3/4) # Screen device commented out
#'                                      # due to CRAN test requirements.
#'                                      # Use it when trying this example
#' xlim <- c(-3, 3)
#' ylim <- ylimFromXlim(xlim, centerY = -0.3, x_to_y = 4/3)
#' op <- par(mar = c(0, 0, 0, 0), bg = "black") # Omit all plot margins
#' phasePortrait(jacobiTheta, moreArgs = list(tau = 1i/2 - 1/3),
#'   xlim = xlim, ylim = ylim, # Apply the coordinate ranges
#'   xaxs = "i", yaxs = "i",   # Allow for now room between plot and axes
#'   nCores = 1) # Max. two cores allowed on CRAN
#'               # not a limit for your own use
#' par(op)
#'   \dontshow{
#'   # R CMD check: make sure any open connections are closed afterward
#'   foreach::registerDoSEQ()
#'   doParallel::stopImplicitCluster()
#'   }
#' }
#'
#'
ylimFromXlim <- function(xlim, centerY = 0, x_to_y = 16/9) {

  yRange <- abs(diff(xlim)) / x_to_y
  ylim   <- centerY + c(-1/2, 1/2) * yRange

  return(ylim)
}



#' Adjust xlim to ylim
#'
#' This simple function is useful for adjusting x and y coordinate ranges
#' \code{xlim} and \code{ylim} in order to maintain a desired display ratio. The
#' latter must be given, the former will be adjusted.
#'
#' For certain purposes, e.g. producing a graph that exactly matches a screen,
#' the x and y coordinates must be adjusted to match a given display ratio. If
#' the vertical range, \code{ylim}, the desired ratio, \code{x_to_y} and the
#' desired center of the x-range, \code{centerX}, are provided, this function
#' returns an adpated vertical range, that can be used as \code{ylim} in any
#' plot including \code{\link{phasePortrait}}.
#'
#' @param ylim Numeric vector of length 2; the fixed lower and upper boundary
#'   of the vertical coordinate range
#'
#' @param centerX The horizontal coordinate which the output range is to be
#'   centered around (default = 0)
#'
#' @param x_to_y The desired ratio of the horizontal (x) to the vertical (y)
#'   range. Default is 16/9, a display ratio frequently used for computer or
#'   mobile screens
#'
#' @return A numeric vector of length 2; the lower and upper boundary of the
#'   resulting vertical coordinate range
#'
#' @family helpers
#'
#' @export
#'
#' @examples
#' # Make a phase portrait of a pretty function that fully covers a
#' # plot with a display aspect ratio of 5/4.
#'
#' # 9 inch wide window with 5/4 display ratio (x/y)
#' \donttest{
#' # x11(width = 9, height = 9 * 4/5) # Screen device commented out
#'                                    # due to CRAN test requirements.
#'                                    # Use it when trying this example
#' ylim <- c(-8, 7)
#' xlim <- xlimFromYlim(ylim, centerX = 0, x_to_y = 5/4)
#' op <- par(mar = c(0, 0, 0, 0), bg = "black") # Omit all plot margins
#' phasePortrait("exp(cosh(1/(z - 2i + 2)^2 * (1/2i - 1/4 + z)^3))", pType = "pm",
#' xlim = xlim, ylim = ylim, # Apply the coordinate ranges
#' xaxs = "i", yaxs = "i",   # Allow for now room between plot and axes
#' nCores = 2) # Max. two cores allowed on CRAN
#'             # not a limit for your own use
#' par(op)
#'   \dontshow{
#'   # R CMD check: make sure any open connections are closed afterward
#'   foreach::registerDoSEQ()
#'   doParallel::stopImplicitCluster()
#'   }
#' }
#'
#'
xlimFromYlim <- function(ylim, centerX = 0, x_to_y = 16/9) {

  xRange <- abs(diff(ylim)) * x_to_y
  xlim   <- centerX + c(-1/2, 1/2) * xRange

  return(xlim)
}





R/Vector2String.R
# R package viscomplexr - phase portraits of functions in the
# complex number plane
# Copyright (C) 2020 Peter Biber
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>



#' Convert a vector into a comma-separated string
#'
#' A simple utility function that transforms any vector into a single character
#' string, where the former vector elements are separated by commas. This is can
#' be useful, in some circumstances, for feeding a series of constant numeric
#' values to \code{\link{phasePortrait}} (see examples). For most applications
#' we recommend, however, to use \code{\link{phasePortrait}}'s parameter
#' \code{moreArgs} instead.
#'
#' @param vec The (usually real or complex valued) vector to be converted.
#'
#' @return A string, where the former vector elements are separated by commas,
#'   enclosed between "c(" and ")".
#'
#' @family helpers
#'
#' @export
#'
#' @examples
#' # Make a vector of 77 complex random numbers inside the unit circle
#' n <- 77
#' a <- complex(n, modulus = runif(n), argument = 2*pi*runif(n))
#' a <- vector2String(a)
#' print(a)
#'
#'
#' # Use this for portraying a Blaschke product
#' \donttest{
#' # x11(width = 9.45, height = 6.30) # Screen device commented out
#'                                    # due to CRAN test requirements.
#'                                    # Use it when trying this example
#' op <- par(mar = c(1, 1, 1, 1), bg = "black")
#' n <- 77
#' a <- complex(n, modulus = runif(n), argument = 2*pi*runif(n))
#' a <- vector2String(a)
#' FUN <- paste("vapply(z, function(z, a){
#'                     return(prod(abs(a)/a * (a-z)/(1-Conj(a)*z)))
#'                    }, a =", a,
#'              ", FUN.VALUE = complex(1))", sep = "")
#' phasePortrait(FUN, pType = "p", axes = FALSE,
#'               xlim = c(-3, 3), ylim = c(-2.0, 2.0),
#'               nCores = 2) # Max. two cores allowed on CRAN
#'                           # not a limit for your own use
#' par(op)
#'   \dontshow{
#'   # R CMD check: make sure any open connections are closed afterward
#'   foreach::registerDoSEQ()
#'   doParallel::stopImplicitCluster()
#'   }
#' }
#'
#'
vector2String <- function(vec) {

  n    <- length(vec)
  rVec <- paste(vec, collapse = ", ")
  rVec <- paste("c(", rVec, ")", sep = "")
  return(rVec)

}


R/RiemannMask.R
# R package viscomplexr - phase portraits of functions in the
# complex number plane
# Copyright (C) 2020 Peter Biber
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>



#' Plot a Riemann sphere mask over a phase portrait
#'
#' The function \code{riemannMask} can be used for laying a circular mask over
#' an existing phasePortrait (as generated with the function
#' \code{\link{phasePortrait}}). This mask shades the plot region outside the
#' unit circle. The unshaded area is a projection on the southern or northern
#' Riemann hemisphere. The standard projection used by
#' \code{\link{phasePortrait}}, i.e. \code{invertFlip = FALSE} hereby
#' corresponds to the southern Riemann hemisphere with the origin being the
#' south pole. If \code{\link{phasePortrait}} was called with \code{invertFlip =
#' TRUE}, then the unit circle contains the northern Riemann hemisphere with the
#' point at infinity in the center (see the vignette for more details). Options
#' for adding annotation, landmark points  are available
#' (\insertCite{@see @wegert_visualcpx_2012;textual}{viscomplexr}, p. 41).
#' Several parameters are on hand for adjusting the mask's transparency, color,
#' and similar features. some details, this function behaves less nicely under
#' Windows than under Linux (see Details).
#'
#' There is, unfortunately, a somewhat different behavior of this function under
#' Linux and Windows systems. Under Windows, the region outside the unit circle
#' is only shaded if the whole unit circle fits into the plot region. If only a
#' part of the unit circle is to be displayed, the shading is completely omitted
#' under Windows (annotation etc. works correctly, however), while it works
#' properly on Linux systems. Obviously, the function \code{\link{polypath}},
#' which we are using for creating the unit circle template, is interpreted
#' differently on both systems.
#'
#' @param colMask Color for the shaded area outside the unit circle. Defaults to
#'   "white". Can be any kind of color definition R accepts. I recommend,
#'   however, to use a color definition without a transparency value, because
#'   this would be overridden by the parameter \code{alphaMask}.
#'
#' @param alphaMask Transparency value for the color defined with
#'   \code{colMask}. Has to be a value between 0 (fully transparent) and 1
#'   (totally opaque). Defaults to 0.5.
#'
#' @param circOutline Boolean - if \code{TRUE}, the outline of the unit circle
#'   is drawn. Defaults to
#'   \code{TRUE}.
#'
#' @param circLwd Line width of the unit circle outline. Obviously relevant
#'   only when \code{circOutline == TRUE}. Defaults to 1.
#'
#' @param circleSteps Number of vertices to draw the circle. Defaults to 360
#'   (one degree between two vertices).
#'
#' @param circleCol Color of the unit circle, default is the default foreground
#'   color (\code{par("fg")}).
#'
#' @param gridCross Boolean - if \code{TRUE}, a horizontal and a vertical gray
#'   line will be drawn over the plot region, intersection in the center of the
#'   unit circle. Defaults to \code{FALSE}.
#'
#' @param annotSouth Boolean - add landmark points and annotation for a
#'   \emph{southern} Riemann hemisphere, defaults to \code{FALSE}. This
#'   annotation fits to an image that has been created with
#'   \code{\link{phasePortrait}} and the option \code{invertFlip = FALSE}.
#'
#' @param annotNorth Boolean - add landmark points and annotation for a
#'   \emph{northern} Riemann hemisphere, defaults to \code{FALSE}. This
#'   annotation fits to an image that has been created with
#'   \code{\link{phasePortrait}} and the option \code{invertFlip = TRUE}.
#'
#' @param xlim,ylim optional, if provided must by numeric vectors of length 2
#'   defining plot limits as usual. They define the outer rectangle of the
#'   Riemann mask. If \code{xlim} or \code{ylim} is not provided (the standard
#'   case), the coordinates of the plot window as given by \code{par("usr")}
#'   will be used for the missing component.
#'
#' @references
#'   \insertAllCited{}
#'
#' @examples
#' # Tangent with fully annotated Riemann masks.
#' # The axis tick marks on the second diagram (Northern hemisphere)
#' # have to be interpreted as the real and imaginary parts of 1/z
#' # (see vignette). The axis labels in this example have been adapted
#' # accordingly.
#' \donttest{
#' # x11(width = 16, height = 8) # Screen device commented out
#'                               # due to CRAN test requirements.
#'                               # Use it when trying this example
#' op <- par(mfrow = c(1, 2), mar = c(4.7, 4.7, 3.5, 3.5))
#' phasePortrait("tan(z)", pType = "pma",
#'               main = "Southern Riemann Hemisphere",
#'               xlim = c(-1.2, 1.2), ylim = c(-1.2, 1.2),
#'               xlab = "real", ylab = "imaginary",
#'               xaxs = "i", yaxs = "i",
#'               nCores = 2) # Max. two cores on CRAN, not a limit for your use
#'   \dontshow{
#'   # R CMD check: make sure any open connections are closed afterward
#'   foreach::registerDoSEQ()
#'   doParallel::stopImplicitCluster()
#'   }
#' riemannMask(annotSouth = TRUE, gridCross = TRUE)
#'
#' phasePortrait("tan(z)", pType = "pma",
#'               main = "Northern Riemann Hemisphere",
#'               invertFlip = TRUE,
#'               xlim = c(-1.2, 1.2), ylim = c(-1.2, 1.2),
#'               xlab = "real (1/z)", ylab = "imaginary (1/z)",
#'               xaxs = "i", yaxs = "i",
#'               nCores = 2) # Max. two cores on CRAN, not a limit for your use
#'   \dontshow{
#'   # R CMD check: make sure any open connections are closed afterward
#'   foreach::registerDoSEQ()
#'   doParallel::stopImplicitCluster()
#'   }
#' riemannMask(annotNorth = TRUE, gridCross = TRUE)
#' par(op)
#' }
#'
#' # Rational function with Riemann masks without annotation.
#' # The axis tick marks on the second diagram (Northern hemisphere)
#' # have to be interpreted as the real and imaginary parts of 1/z
#' # (see vignette). The axis labels in this example have been adapted
#' # accordingly.
#' \donttest{
#' # x11(width = 16, height = 8) # Screen device commented out
#'                               # due to CRAN test requirements.
#'                               # Use it when trying this example
#' op <- par(mfrow = c(1, 2), mar = c(4.7, 4.7, 3.5, 3.5))
#' phasePortrait("(-z^17 - z^15 - z^9 - z^7 - z^2 - z + 1)/(1i*z - 1)",
#'               pType = "pma",
#'               main = "Southern Riemann Hemisphere",
#'               xlim = c(-1.2, 1.2), ylim = c(-1.2, 1.2),
#'               xlab = "real", ylab = "imaginary",
#'               xaxs = "i", yaxs = "i",
#'               nCores = 2) # Max. two cores on CRAN, not a limit for your use
#'   \dontshow{
#'   # R CMD check: make sure any open connections are closed afterward
#'   foreach::registerDoSEQ()
#'   doParallel::stopImplicitCluster()
#'   }
#' riemannMask(annotSouth = FALSE, gridCross = FALSE, circOutline = FALSE,
#'             alphaMask = 0.7)
#'
#' phasePortrait("(-z^17 - z^15 - z^9 - z^7 - z^2 - z + 1)/(1i*z - 1)",
#'               pType = "pma",
#'               main = "Northern Riemann Hemisphere",
#'               invertFlip = TRUE,
#'               xlim = c(-1.2, 1.2), ylim = c(-1.2, 1.2),
#'               xlab = "real (1/z)", ylab = "imaginary (1/z)",
#'               xaxs = "i", yaxs = "i",
#'               nCores = 2) # Max. two cores on CRAN, not a limit for your use
#'   \dontshow{
#'   # R CMD check: make sure any open connections are closed afterward
#'   foreach::registerDoSEQ()
#'   doParallel::stopImplicitCluster()
#'   }
#' riemannMask(annotNorth = FALSE, gridCross = FALSE, circOutline = FALSE,
#'             alphaMask = 0.7)
#' par(op)
#' }
#'
#'
#' @export
#'
#'
riemannMask <- function(colMask     = "white",
                        alphaMask   = 0.5,
                        circOutline = TRUE,
                        circLwd     = 1,
                        circleSteps = 360,
                        circleCol   = par("fg"),
                        gridCross   = FALSE,
                        annotSouth  = FALSE,
                        annotNorth  = FALSE,
                        xlim        = NULL,
                        ylim        = NULL) {

  # Get user plot coordinate extremes and calculate
  # the widths in both directions (x and y)
  coord  <- par("usr")
  xlmt   <- c(coord[1],  coord[2])
  ylmt   <- c(coord[3],  coord[4])

  # Overwrite if xlim and/or ylim are not NULL
  if(!is.null(xlim)) xlmt <- xlim
  if(!is.null(ylim)) ylmt <- ylim

  # Define the outer frame of the mask
  frame  <- list(x = c(xlmt[1], xlmt[1], xlmt[2], xlmt[2]),
                 y = c(ylmt[1], ylmt[2], ylmt[2], ylmt[1]))

  # Define the inner unit circle
  innerCircle <- list(x = cos(seq(360/circleSteps, 360,
                                  by = 360/circleSteps)*2*pi/360),
                      y = sin(seq(360/circleSteps, 360,
                                  by = 360/circleSteps)*2*pi/360))

  # Plot it
  polypath(x = c(frame$x, NA, innerCircle$x),
           y = c(frame$y, NA, innerCircle$y),
           col  = scales::alpha(colMask, alpha = alphaMask),
           rule = "evenodd",
           border = NA)

  # Add circle
  if(circOutline) {
    plotrix::draw.circle(0, 0, 1, nv = circleSteps,
                         lwd = circLwd, border = circleCol)
  }

  # Add grid cross if desired
  if(gridCross) {
    abline(h = 0, col = "grey")
    abline(v = 0, col = "grey")
  }

  # Add annotation if desired (for southern hemisphere)
  if(annotSouth) {
    pts <- list(x = c(-1, 0, 0, 0, 1), y = c(0, 1, 0, -1, 0))
    points(pts$x, pts$y, pch = 21, col = "black",
           bg = c("white", "white", "black", "white", "white"))

    labels <- c("-1", "i", "0", "-i", "1")
    i <- c(1:5)
    lapply(i, function(i, pts, labels) {
      adx    <- switch(labels[i],
        "-1" =   1.5, "i" =  -2.5, "0" = -2.0, "-i" = -0.6, "1" = -1.8)
      ady    <- switch(labels[i],
        "-1" =  -1.2, "i" =  -1.2, "0" = -1.2, "-i" =  2.2, "1" = -1.2)
      text(pts$x[i], pts$y[i], labels[i], vfont = c("serif", "bold"),
           adj = c(adx, ady))
    }, pts = pts, labels = labels)
  } # if annotSouth

  # Add annotation if desired (for northern hemisphere)
  if(annotNorth) {
    pts <- list(x = c(-1, 0, 0, 0, 1), y = c(0, 1, 0, -1, 0))
    points(pts$x, pts$y, pch = 21, col = "black",
           bg = c("white"))

    labels <- c("-1", "i", " ", "-i", "1")
    i <- c(1:5)
    lapply(i, function(i, pts, labels) {
      adx    <- switch(labels[i],
        "-1" =  -0.6, "i" =  -2.5, " " = -2.0, "-i" = -0.6, "1" =  2.6)
      ady    <- switch(labels[i],
        "-1" =  -1.2, "i" =  -1.2, " " = -1.2, "-i" =  2.2, "1" = -1.2)
      text(pts$x[i], pts$y[i], labels[i], vfont = c("serif", "bold"),
           adj = c(adx, ady))
    }, pts = pts, labels = labels)

    text(0, 0, expression(infinity), adj = c(-0.7, -1.0), cex = 1.7, font = 2)
  } # if annotNorth

} # function riemannMask

# ------------------------------------------------------------------------------
# ------------------------------------------------------------------------------

